Profile management

ctrlaltca opened the pull request:
This PR adds 3 buttons in the "account" tab:
 * edit user information
 * change password
 * change avatar

![account userinfobox](https://cloud.githubusercontent.com/assets/1631111/8512109/f5ea6d28-2334-11e5-81f4-e827e410a750.png)

Each button opens a new dialog:

![edit user](https://cloud.githubusercontent.com/assets/1631111/8512117/0ca2d712-2335-11e5-87fe-55d795ccf813.png)
![change pass](https://cloud.githubusercontent.com/assets/1631111/8512115/0ca225e2-2335-11e5-9325-c1ac54d97fdf.png)
![change avatar](https://cloud.githubusercontent.com/assets/1631111/8512116/0ca27fce-2335-11e5-9688-83c6f3d3ae00.png)

poixen commented:
This is awesome. Makes me wish we waited on a new release.

Daenyth commented on cockatrice/src/dlg_edit_avatar.cpp:
```diff
@@ -0,0 +1,89 @@
+    imageLabel->setAlignment(Qt::AlignHCenter | Qt::AlignVCenter);
+    imageLabel->setStyleSheet("border: 1px solid #000");
+
+    textLabel = new QLabel(tr("To change your avatar, choose a new image.\nTo remove your current avatar, just confirm without choosing a new image."));
```
I'd leave out `just`

Daenyth commented on cockatrice/src/dlg_edit_avatar.cpp:
```diff
@@ -0,0 +1,89 @@
+    QImageReader imgReader;
+    imgReader.setDecideFormatFromContent(true);
+    imgReader.setFileName(fileName);
+    if(!imgReader.read(&image))
```
Can you debug log any errors from this?

Daenyth commented on cockatrice/src/dlg_edit_password.cpp:
```diff
@@ -0,0 +1,60 @@
+    oldPasswordLabel->setBuddy(oldPasswordEdit);
+    oldPasswordEdit->setEchoMode(QLineEdit::Password);
+
+    newPasswordLabel = new QLabel(tr("New password:"));
```
Can you make a second field so that the user is forced to type twice and confirm that the password is correct?

Daenyth commented on cockatrice/src/dlg_edit_password.cpp:
```diff
@@ -0,0 +1,60 @@
+{
+    QSettings settings;
+    settings.beginGroup("server");
+    // always save the password so it will be picked up by the connect dialog
```
We should honor the "Save password" config though, it's not always enabled.

Daenyth commented on cockatrice/src/dlg_edit_user.cpp:
```diff
@@ -0,0 +1,326 @@
+    emailLabel->setBuddy(emailEdit);
+    emailEdit->setText(email);
+
+    genderLabel = new QLabel(tr("Gender:"));
```
Can you label this as pronouns, neutral, masculine, feminine?

ZeldaZach replied:
Seems to have been done already

Daenyth replied:
I was also saying "Gender" should be changed to "Pronouns"

ctrlaltca replied:
Opened #1224

Daenyth commented on cockatrice/src/dlg_edit_user.cpp:
```diff
@@ -0,0 +1,326 @@
+    countryEdit = new QComboBox();
+    countryLabel->setBuddy(countryEdit);
+    countryEdit->insertItem(0, tr("Undefined"));
+    countryEdit->addItem(QPixmap(":/resources/countries/ad.svg"), "ad");
```
Can we take a glob? The repetition here seems a little excessive.

Daenyth commented on cockatrice/src/dlg_edit_user.cpp:
```diff
@@ -0,0 +1,326 @@
+    mainLayout->addWidget(buttonBox);
+    setLayout(mainLayout);
+
+    setWindowTitle(tr("Edit user informations"));
```
"Edit user profile"?

ZeldaZach commented:
@poixen it will be in the next weekly build and we have revisions of Lightning Angel; we're on revision 1 atm

Great job @ctrlaltca I'll do some testing later tonight :)

Daenyth commented on cockatrice/src/userinfobox.cpp:
```diff
@@ -140,3 +162,126 @@ void UserInfoBox::processResponse(const Response &r)
+
+void UserInfoBox::processEditResponse(const Response &r)
+{
+    switch (r.response_code()) {
```
We should check RespInvalidCommand, I think that will get sent if the server is too old

ctrlaltca replied:
If the server is too old it should send Response::RespFunctionNotAllowed (the related code is in ServerSocketInterface::processExtendedSessionCommand() )

Daenyth replied:
:+1:

Daenyth commented on cockatrice/src/userinfobox.cpp:
```diff
@@ -140,3 +162,126 @@ void UserInfoBox::processResponse(const Response &r)
+        case Response::RespFunctionNotAllowed:
+            QMessageBox::critical(this, tr("Error"), tr("This server does not permit you to change your password."));
+            break;
+        case Response::RespPasswordTooShort:
```
What if it fails other username validation rules? We should ideally use the logic from login

ctrlaltca replied:
We are not validating the username here, only the new password. The only (hardcoded) rule by now is length >= 6 chars

Daenyth replied:
derp nevermind

Daenyth commented on cockatrice/src/userinfobox.cpp:
```diff
@@ -140,3 +162,126 @@ void UserInfoBox::processResponse(const Response &r)
+
+void UserInfoBox::processAvatarResponse(const Response &r)
+{
+    switch (r.response_code()) {
```
We should have a response for size limits, and the server should be able to configure the size limit

ctrlaltca replied:
The database is using a BLOB to store the image, than in mysql means 64kbs.. probably not enough for a big image. That's why currently the user-chosen image is resized down to 400x200 and sent in jpg format.
I meant to address this making mysql use a MEDIUMBLOB (16mbs) and adding the relevant server-side checks and settings.
This would touch some more files and add more code, so i was planning on fixing it on a following PR.

Daenyth replied:
+1 for another pr

Daenyth commented on servatrice/src/servatrice_database_interface.cpp:
```diff
@@ -471,6 +471,10 @@ ServerInfo_User Servatrice_DatabaseInterface::evalUserQueryResult(const QSqlQuer
             result.set_accountage_secs(accountAgeInSeconds);
         }
+
+        const QString email = query->value(8).toString();
```
Not needed for this PR, but we should make an enum for the sql column fields

Daenyth commented on servatrice/src/servatrice_database_interface.cpp:
```diff
@@ -756,4 +760,41 @@ void Servatrice_DatabaseInterface::logMessage(const int senderId, const QString
+    if (!usernameIsValid(user))
+        return true;
+
+    QSqlQuery *passwordQuery = prepareQuery("select password_sha512 from {prefix}_users where name = :name");
```
Is the query deleted by some other function? Will this leak?

ctrlaltca replied:
The prepareQuery() method will create a sql prepared statement and save it in a cache. These prepared statemen are never freed, but will be reused when needed.

Daenyth replied:
Right I forgot

ctrlaltca commented:
I've addressed the issues, except these problems:
 * the avatar image size is too small in the db and not checked anywhere;
 * the password constraints are not configurable on the server
 * travis seems to be failing at compiling this, or i've broken something with git.

Daenyth commented:
```
/home/travis/build/Cockatrice/Cockatrice/servatrice/src/servatrice_database_interface.cpp: In member function ‘virtual bool Servatrice_DatabaseInterface::changeUserPassword(const QString&, const QString&, const QString&)’:
/home/travis/build/Cockatrice/Cockatrice/servatrice/src/servatrice_database_interface.cpp:778:30: error: no matching function for call to ‘Servatrice_DatabaseInterface::usernameIsValid(const QString&)’
/home/travis/build/Cockatrice/Cockatrice/servatrice/src/servatrice_database_interface.cpp:778:30: note: candidate is:
/home/travis/build/Cockatrice/Cockatrice/servatrice/src/servatrice_database_interface.cpp:121:6: note: virtual bool Servatrice_DatabaseInterface::usernameIsValid(const QString&, QString&)
/home/travis/build/Cockatrice/Cockatrice/servatrice/src/servatrice_database_interface.cpp:121:6: note:   candidate expects 2 arguments, 1 provided
```

ctrlaltca commented:
Rebased on master, will hopefully fix compilation

woogerboy21 commented:
gcc's still crying on this one?
```
/home/travis/build/Cockatrice/Cockatrice/build/cockatrice/qrc_cockatrice.cpp: In function ‘int qInitResources_cockatrice()’:
/home/travis/build/Cockatrice/Cockatrice/build/cockatrice/qrc_cockatrice.cpp:195660:5: warning: no previous declaration for ‘int qInitResources_cockatrice()’ [-Wmissing-declarations]
/home/travis/build/Cockatrice/Cockatrice/build/cockatrice/qrc_cockatrice.cpp: In function ‘int qCleanupResources_cockatrice()’:
/home/travis/build/Cockatrice/Cockatrice/build/cockatrice/qrc_cockatrice.cpp:195669:5: warning: no previous declaration for ‘int qCleanupResources_cockatrice()’ [-Wmissing-declarations]

```

ctrlaltca commented:
That warning is "normal" and is caused by qt resource system. The problem was gcc complaining some lines before for unused variables.

woogerboy21 commented:
Oh ok cool... This is awesome :+1:

woogerboy21 commented:
@ctrlaltca : so are the function calls for updating the profile data written to easily allow an administration portion inside the app to be created which will allow admin/moderators the ability to manipulate user accounts?  Its pretty common that users some times need there passwords reset or accounts activated for example.

woogerboy21 commented:
FYI.  This worked on the testing server with no problems :+1: 
:shipit:

ZeldaZach commented:
@woogerboy21 Can you briefly explain how you tested everything? Just curious as I want to do my own tests

woogerboy21 commented:
Roll the server and client.  Log in with a registered account.  Go to your account tab.  Click the buttons accordingly.

ctrlaltca commented:
@woogerboy21 a new "profile administration" functionality will require new admin-only commands, but they'll mostly be clones of the session commands added in this PR. The gui is mostly adaptable.

ZeldaZach commented:
Seems good @ctrlaltca. If you can do the debugging @Daenyth requested, we'll be ready to merge this in :smile:
